src/server/http/xsrf.js

```diff
@@ -4,18 +4,48 @@ export default function (kbnServer, server, config) {
   const disabled = config.get('server.xsrf.disableProtection');
   const versionHeader = 'kbn-version';
   const xsrfHeader = 'kbn-xsrf';
+  const contentTypeHeader = 'content-type';
+  const allowedRequestMediaTypes = ['application/json', 'application/x-ndjson'];
 
   server.ext('onPostAuth', function (req, reply) {
     if (disabled) {
       return reply.continue();
     }
 
     const isSafeMethod = req.method === 'get' || req.method === 'head';
-    const hasVersionHeader = versionHeader in req.headers;
-    const hasXsrfHeader = xsrfHeader in req.headers;
+    if (!isSafeMethod) {
+      const hasVersionHeader = versionHeader in req.headers;
+      const hasXsrfHeader = xsrfHeader in req.headers;
 
-    if (!isSafeMethod && !hasVersionHeader && !hasXsrfHeader) {
-      return reply(badRequest(`Request must contain an ${xsrfHeader} header`));
+      if (hasXsrfHeader) {
+        server.log(
+          ['warning', 'deprecation'],
+          `The ${xsrfHeader} header is deprecated and will be removed in a future version of Kibana.` +
+          ` Specify a ${contentTypeHeader} header of either application/json or application/x-ndjson instead.`
```

Comment:
Re "test-verbosity" I'm not actually sure if we have a clear guideline, maybe @epixa has some good wording around that. I just test to the level where I feel confident in the feature and that it's obvious for the next person reading the code+test what it's supposed to do. (I realize that is very subjective though ;))

The reason I'm asking for more tests here is that this is an important feature to get right and the logs are a part of the feature itself (for most features the logging itself isn't necessarily part of the feature, but for deprecations I think they are important). Also, here we have three different fields that interplay, so it's easy to break the feature by moving code around.

I'm not sure if there are examples in the codebase, and it might not have been done before. /cc @epixa